### DISCUSSION TITLE
Function .end

### DISCUSSION THREAD

**Discussion by BluePython339** [Vector35/binaryninja-api#2189] [API] [answered]
The class function has a `.start` attribute, yet not a `.end`, i'm trying to calculate the amount of instructions inside the function without using the full generator, i tried to do that using a simple end - start but got the error that there was not `.end`. Why is this, and how can i solve it

**Comment by plafosse** [comment, marked as answer]
End as a concept doesn't really make sense in the way Binary Ninja defines a Function (a collection of connected BasicBlock objects).  Consider the following code.
sub_0:
 call sub_2
 ret

b:
 ret
sub_2:
 jmp b

In the above function there is no clear "end" of the function. We feel that if we created such an API it would be error prone as people have different ideas on how the function "end" should be calculated.

**Reply by psifertex** [reply]
To summarize, there are many reasons why we don't include `.end`. The main one is that there's no clear definition even as agreed upon by other reverse engineering tools! Other tools all handle edge-cases differently. What about multiple return functions? What about functions where the return isn't the highest address? Each of those can cause different behaviors and hiding that complexity behind `.end` is a bad idea. 

We do, however, provide the `.total_bytes` property which is the sum of the sizes of basic_blocks. Because basic_blocks are well-defined with a start and an end and functions are explicitly a collection of basic blocks. There is also the `.highest_address` and `.lowest_address` which could be subtracted to compute a range of total bytes as well. Those numbers are not necessarily the same though for multiple reasons: functions may be dis-contiguous, basic blocks can overlap, or other reasons.
